lgc/patch/PatchWorkarounds.cpp

```diff
@@ -0,0 +1,175 @@
+// Process calls to image instrinsics and apply buffer descriptor should be image descriptor workaround
+//
+// @param callInst : The image intrinsic call instruction
+void PatchWorkarounds::processImageIntrinsic(CallInst &callInst) {
+  Function *const calledFunc = callInst.getCalledFunction();
+  if (!calledFunc)
+    return;
+
+  // A buffer descriptor may be incorrectly given when it should be an image descriptor, we need to fix it to valid
+  // buffer type (0) to make hardware happily ignore it. This is to check and fix against buggy applications which
+  // declares a image descriptor in shader but provide a buffer descriptor in driver. Note this only applies to gfx10.1.
+  // Look for 8 dword resource descriptor argument
+  for (Value *arg : callInst.args()) {
+    if (auto vecTy = dyn_cast<FixedVectorType>(arg->getType())) {
+      if (vecTy->getNumElements() == 8 && vecTy->getElementType()->isIntegerTy(32)) {
+        if (isa<UndefValue>(arg))
+          // We don't need to worry if the value is actually undef. This situation only really occurs in unit test
+          // but either way, it is pointless to apply the workaround to an undef.
+          break;
+
+        if (m_processed.count(arg) > 0)
+          // Already processed this one
+          break;
+
+        Instruction *rsrcInstr = cast<Instruction>(arg);
+
+        m_builder->SetInsertPoint(rsrcInstr->getNextNode());
+        m_builder->SetCurrentDebugLocation(rsrcInstr->getNextNode()->getDebugLoc());
+
+        // Create a new rsrc load instruction - we apply the workaround the the new instruction and then replace
+        // all uses of the old one with the derived value. This prevents us replacing the original value with the
+        // derived one in the code to insert the new element.
+        Instruction *newInstr = rsrcInstr->clone();
+        newInstr->insertAfter(rsrcInstr);
+
+        Value *elem3 = m_builder->CreateExtractElement(newInstr, 3);
+        Value *isInvalid = m_builder->CreateICmpSGE(elem3, m_builder->getInt32(0));
+        Value *masked = m_builder->CreateAnd(elem3, m_builder->getInt32(0x0FFFFFFF));
+        elem3 = m_builder->CreateSelect(isInvalid, masked, elem3);
+
+        // Re-assemble descriptor
+        Value *newArg = m_builder->CreateInsertElement(newInstr, elem3, 3);
+
+        auto rsrcInstrName = rsrcInstr->getName();
+        rsrcInstr->replaceAllUsesWith(newArg);
+        rsrcInstr->eraseFromParent();
+        newInstr->setName(rsrcInstrName); // Preserve the old name for the load
+        newArg->setName(rsrcInstrName);   // Derive a new name based on the old name for the load
+
+        // Record the new argument as already processed. If we encounter it in
+        // another image intrinsic call we can skip
+        m_processed.insert(newArg);
+
+        m_changed = true;
+        break;
+      }
+    }
+  }
+}
+
+void PatchWorkarounds::getAnalysisUsage(llvm::AnalysisUsage &analysisUsage) const {
```

Comment:
Please add function header comments for getAnalysisUsage. Pleased remove "llvm::" prefix in .cpp files.

Reply:
About to upload a patch that does this - waiting to get to agreement on next question